### DISCUSSION TITLE
How to sync trpc websockets / events over multiple servers

### DISCUSSION THREAD

**Discussion by xzilja** [trpc/trpc#3423] [Q&A] [answered]
Hey 👋 

Had a question regarding trpc usage as a websocket api that is distributed across multiple servers i.e in horizontal scaling scenarios. There could be a case when 2 users are connected to 2 different server instances, but are both subscribed to one topic.
If topic is emitted on one server, how can we ensure second subscriber (on a diff server) will receive it?

In ws servers like socket.io this is solved via "adapters", with most popular being redis one https://socket.io/docs/v4/redis-adapter

Is something like this doable via trpc?

**Comment by KATT** [comment, marked as answer]
Yeah, you need to subscribe to something like Redis within your subscription resolvers 


https://github.com/trpc/trpc/blob/444bbd6bdb314570a0f1ecc0cf646c2cb64b26fd/examples/next-prisma-websockets-starter/src/server/routers/post.ts#L28-L29

**Reply by Blankeos** [reply]
Question, do I still use the node event emitter class or do I just use Redis itself?

I'm wondering how Redis (a key value store) replaces an Event Emitter (pubsub). Any examples?

Or do I build my own pubsub using Redis?

**Reply by BeBoRE** [reply]
Redis supports pubsub
https://redis.io/docs/latest/develop/interact/pubsub/
https://www.npmjs.com/package/ioredis#pubsub
And then you can publish messages in your mutations and listen for those in your subscriptions

> Also since this post was made almost 2 years ago, it's probably better to use Valkey, because of you [know](https://redis.io/blog/redis-adopts-dual-source-available-licensing/).
